// couponscom-util/Gruntfile.js
'use strict';

module.exports = function (grunt) {

    grunt.initConfig({
        lint: {
            files: ['*.js', 'test/*.js'],
            options: {
                jshintrc: '../.jshintrc',
                reporter: 'jslint',
                reporterOutput: './jshint-results.xml'
            }
        },
        unittest: {
            src: ['test/*.js'],
            options: {
                globals: ['chai'],
                timeout: 6000,
                ignoreLeaks: false,
                ui: 'bdd',
                reporter: 'mocha-jenkins-reporter'
            }
        },
        shell: {
            'coverage': {
                options: {
                    stdout: true
                },
                command: './node_modules/istanbul/lib/cli.js cover ./node_modules/grunt-mocha-cli/node_modules/mocha/bin/_mocha -- --ui bdd -t 5000'
            }
        }
    });

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-mocha-cli');
    grunt.loadNpmTasks('grunt-shell');

    grunt.renameTask('mochacli', 'unittest');
    grunt.renameTask('jshint', 'lint');

    grunt.registerTask('test', ['lint', 'unittest', 'shell:coverage']);
};
